### DISCUSSION TITLE
Initial Condition as an array instead of function

### DISCUSSION THREAD

**Discussion by ryabs07** [lululxvi/deepxde#1419] [Q&A] [answered]
Hi all,
Consider a domain x ranging from 0 to 1. I want to give initial conditions for a variable 'u' within this domain which cannot be described by a function of x, nor does it have a constant value. I have the initial values of u in array that is descretized to 500 points uniformly over the space domain. How do I implement this kind of initial condition problem in DeepXDE?
Thank you.

**Comment by nirmal129** [comment, marked as answer]
Hello, 

I believe PointSetBC will help.

**Reply by ryabs07** [reply]
Hi Nirmal,
Thank you for the reply.  I suppose this is the function I was looking for. Although it is called PointSet**BC**, I wonder if it can be used for initial condition as well.
Anyways, I tried the following for my problem statement:

```
x = np.linspace(x_start, x_end, total_points)
X, T_initial = np.meshgrid(x,0)  #creating a mesh of discretized points and initial time 0
IC_input = np.hstack((X.flatten()[:,None], T_initial.flatten()[:,None]))

#Define the PDEs
def pde(a, b):
    #a is a 2D input array where a[:,0:1] represents spatial domain in x direction and a[:,1:2] represents the temporal domain
    #a is the input of the neural net
    #b is a the output array where y[:,0:1] represents water depth , y[:,1:2] represents velocity, y[:,2:3] represents bed elevation (topography)
    #b is the output of the neural net
    x, t = a[:, 0:1], a[:, 1:2]
    h, u, z = b[:, 0:1], b[:, 1:2], b[:, 2:3]
    ...defining rest of the equations...


ic_h = dde.icbc.PointSetBC(IC_input, h_initial, component=0) #initial water depth as an array
ic_u = dde.icbc.PointSetBC(IC_input, u_initial, component=1) #initial water velocity as an array
ic_z = dde.icbc.PointSetBC(IC_input, z_initial, component=2) #initial topography as an array

data = dde.data.TimePDE(geomtime, pde, [ic_h, ic_u, ic_z, bc_l_h, bc_r_h, bc_l_u], num_domain=20000, num_boundary=1000, anchors = input)
```

However, the loss on these three initial conditions defined by PointSetBC  stops decreasing after a couple of hundreds of iterations. I have tried multiple combination of weights as well. But nothing seems to work. If you have any experience and idea related to this, please let me know. 
Thank you again for your comment.

-Rabs

**Reply by nirmal129** [reply]
Hello Rabs, 

It's good to hear that I was able to help you. I do believe that any BC subclass can be used for any input coordinate. 

Regarding the NN training, I am just a beginner so I don't have any proper solution but there are a bunch of things you can explore apart from weights like input scaling, output transformation, changing the depth/width of the network, using different learning rates, etc. 

Sincerely, 
Nirmal
